File: kotlin-typescript/karakum/annotations/annotateVarOverrides.js
import ts from "typescript"

export default (node) => {
    if (
        ts.isPropertySignature(node)
        && ts.isIdentifier(node.name)
        && (
            node.name.text === "kind"
            || node.name.text === "type"
        )

        && ts.isInterfaceDeclaration(node.parent)
        && (
            node.parent.name.text === "AssertsIdentifierTypePredicate"
            || node.parent.name.text === "AssertsThisTypePredicate"
            || node.parent.name.text === "IdentifierTypePredicate"
            || node.parent.name.text === "ThisTypePredicate"
        )
    ) {
        return `@Suppress("VAR_TYPE_MISMATCH_ON_OVERRIDE")`
    }

    if (
        ts.isPropertySignature(node)
        && ts.isIdentifier(node.name)
        && node.name.text === "value"

        && ts.isInterfaceDeclaration(node.parent)
        && (
            node.parent.name.text === "BigIntLiteralType"
            || node.parent.name.text === "StringLiteralType"
            || node.parent.name.text === "NumberLiteralType"
        )
    ) {
        return `@Suppress("VAR_TYPE_MISMATCH_ON_OVERRIDE")`
    }

    if (
        ts.isPropertySignature(node)
        && ts.isIdentifier(node.name)
        && node.name.text === "moduleSpecifier"

        && ts.isInterfaceDeclaration(node.parent)
        && node.parent.name.text === "CompletionEntryDataResolved"
    ) {
        return `@Suppress("VAR_TYPE_MISMATCH_ON_OVERRIDE")`
    }

    if (
        ts.isPropertySignature(node)
        && ts.isIdentifier(node.name)
        && node.name.text === "exportMapKey"

        && ts.isInterfaceDeclaration(node.parent)
        && node.parent.name.text === "CompletionEntryDataUnresolved"
    ) {
        return `@Suppress("VAR_TYPE_MISMATCH_ON_OVERRIDE")`
    }

    if (
        ts.isPropertySignature(node)
        && ts.isIdentifier(node.name)
        && (
            node.name.text === "file"
            || node.name.text === "start"
            || node.name.text === "length"
        )

        && ts.isInterfaceDeclaration(node.parent)
        && node.parent.name.text === "DiagnosticWithLocation"
    ) {
        return `@Suppress("VAR_TYPE_MISMATCH_ON_OVERRIDE")`
    }

    return null
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/annotations/annotateVarOverrides.kt
package typescript.karakum.annotations

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.AnnotationContext
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.Node
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isMethodSignature
import typescript.isPropertySignature

fun annotateVarOverrides(node: Node, context: AnnotationContext) = nullable {
    nullable {
        ensure(isPropertySignature(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(
            name.text == "kind"
                    || name.text == "type"
        )

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(
            interfaceNode.name.text == "AssertsIdentifierTypePredicate"
                    || interfaceNode.name.text == "AssertsThisTypePredicate"
                    || interfaceNode.name.text == "IdentifierTypePredicate"
                    || interfaceNode.name.text == "ThisTypePredicate"
        )

        "@Suppress(\"VAR_TYPE_MISMATCH_ON_OVERRIDE\")"
    } ?: nullable {
        ensure(isPropertySignature(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(name.text == "value")

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(
            interfaceNode.name.text == "BigIntLiteralType"
                    || interfaceNode.name.text == "StringLiteralType"
                    || interfaceNode.name.text == "NumberLiteralType"
        )

        "@Suppress(\"VAR_TYPE_MISMATCH_ON_OVERRIDE\")"
    } ?: nullable {
        ensure(isPropertySignature(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(name.text == "moduleSpecifier")

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "CompletionEntryDataResolved")

        "@Suppress(\"VAR_TYPE_MISMATCH_ON_OVERRIDE\")"
    } ?: nullable {
        ensure(isPropertySignature(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(name.text == "exportMapKey")

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "CompletionEntryDataUnresolved")

        "@Suppress(\"VAR_TYPE_MISMATCH_ON_OVERRIDE\")"
    } ?: nullable {
        ensure(isPropertySignature(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(
            name.text == "file"
                    || name.text == "start"
                    || name.text == "length"
        )

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "DiagnosticWithLocation")

        "@Suppress(\"VAR_TYPE_MISMATCH_ON_OVERRIDE\")"
    } ?: nullable {
        ensure(isPropertySignature(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(name.text == "useCaseSensitiveFileNames")

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "ParseConfigHost")

        "@Suppress(\"VAR_TYPE_MISMATCH_ON_OVERRIDE\")"
    } ?: nullable {
        ensure(isPropertySignature(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(name.text == "target")

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "TupleTypeReference")

        "@Suppress(\"VAR_TYPE_MISMATCH_ON_OVERRIDE\")"
    } ?: nullable {
        ensure(isMethodSignature(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(name.text == "useCaseSensitiveFileNames")

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "LanguageServiceHost")

        "@Suppress(\"PROPERTY_TYPE_MISMATCH_ON_OVERRIDE\", \"VAR_OVERRIDDEN_BY_VAL\")"
    }
}
